core/builtin.py

```diff
@@ -235,6 +236,84 @@ def Echo(argv):
   return 0
 
 
+PRINTF_SPEC = _Register('printf')
+PRINTF_SPEC.ShortFlag('-v', args.Str)
+
+
+def Printf(argv, mem):
+  """
+  printf: printf [-v var] format [argument ...]
+  """
+  arg, args_consumed = PRINTF_SPEC.Parse(argv)
+  if args_consumed >= len(argv):
+    util.error('printf: need format string')
+    return 1
+  fmt = argv[args_consumed]
+  vals = argv[args_consumed + 1:]
+
+  result = ''
+  f = 0
+  v = 0
+  # Loop invariant: vals[v:] and fmt[f:] remain, to accumulate into result.
+  while True:
+    f_next = fmt.find('%', f)
+    if f_next < 0:
+      f_next = len(fmt)
+
+    result += fmt[f:f_next]  # TODO backslash-escapes, at least \n
```

Comment:
I think the requirement to lex `\` in addition to `%` means we should use re2c to lex.  See `osh/lex.py PS1_DEF` for a lexer definition for `$PS1`.  It's just a list of regex -> token mappings.

This change could be OK without `\` if it gets us past some practical problems... but long-term I see it being rewritten to use the regex-style lexers.  They are fast because it compiles to C code.

It also tends to help with corner cases, i.e. this style of string processing tends to end up with a lot of IndexError (handled or not).   Using a sentinel (i.e. NUL) tends to be cleaner in my experience (and that's the style that OSH uses throughout).

Reply:
Cool, thanks for the pointer! Took a quick look, and yeah I definitely agree that that should be a better way for this to ultimately be implemented.

As it is, this code is basically C (in my head) translated to Python. I figured a real lexer would be better, but wasn't sure how to get my hands on that in the context of this codebase, and this seemed like a quicker way to get something working in order to see what bash-completion would need just next. C generated from regexes would be much better than Python translated from C :slightly_smiling_face: